chain/bitcoind.go

```diff
@@ -0,0 +1,1122 @@
+	select {
+	case c.rescanUpdate <- txids:
+	case <-c.quit:
+	}
+	return nil
+}
+
+// NotifyBlocks is always on.
+func (c *BitcoindClient) NotifyBlocks() error {
+	return nil
+}
+
+// LoadTxFilter updates the transaction watchlists for the client. Acceptable
+// arguments after `reset` are any combination of []btcutil.Address,
+// []wire.OutPoint, []*wire.OutPoint, []chainhash.Hash, and []*chainhash.Hash.
+func (c *BitcoindClient) LoadTxFilter(reset bool,
+	watchLists ...interface{}) error {
+
+	// If we reset, signal that.
+	if reset {
+		select {
+		case c.rescanUpdate <- reset:
+		case <-c.quit:
+			return nil
+		}
+	}
+
+	// This helper function will send an update to the filter. If the quit
+	// channel is closed, it will allow the outer loop below to finish,
+	// but skip over any updates as the quit case is triggered each time.
+	sendList := func(list interface{}) {
+		select {
+		case c.rescanUpdate <- list:
+		case <-c.quit:
+		}
+	}
+
+	for _, watchList := range watchLists {
+		switch list := watchList.(type) {
+		case []wire.OutPoint:
+			sendList(list)
+		case []*wire.OutPoint:
+			sendList(list)
+		case []btcutil.Address:
+			sendList(list)
+		case []chainhash.Hash:
+			sendList(list)
+		case []*chainhash.Hash:
+			sendList(list)
+		default:
+			log.Warnf("Couldn't add item to filter: unknown type")
+		}
+	}
+	return nil
+}
+
+// RescanBlocks rescans any blocks passed, returning only the blocks that
+// matched as []btcjson.BlockDetails.
+func (c *BitcoindClient) RescanBlocks(blockHashes []chainhash.Hash) (
+	[]btcjson.RescannedBlock, error) {
+
+	rescannedBlocks := make([]btcjson.RescannedBlock, 0, len(blockHashes))
+	for _, hash := range blockHashes {
+		header, err := c.GetBlockHeaderVerbose(&hash)
+		if err != nil {
+			log.Warnf("Unable to get header %s from bitcoind: %s",
+				hash, err)
+			continue
+		}
+
+		block, err := c.GetBlock(&hash)
+		if err != nil {
+			log.Warnf("Unable to get block %s from bitcoind: %s",
+				hash, err)
+			continue
+		}
+
+		relevantTxes, err := c.filterBlock(block, header.Height,
+			false)
+		if len(relevantTxes) > 0 {
+			rescannedBlock := btcjson.RescannedBlock{
+				Hash: hash.String(),
+			}
+			for _, tx := range relevantTxes {
+				rescannedBlock.Transactions = append(
+					rescannedBlock.Transactions,
+					hex.EncodeToString(tx.SerializedTx),
+				)
+			}
+			rescannedBlocks = append(rescannedBlocks,
+				rescannedBlock)
+		}
+	}
+	return rescannedBlocks, nil
+}
+
+// Rescan rescans from the block with the given hash until the current block,
+// after adding the passed addresses and outpoints to the client's watch list.
+func (c *BitcoindClient) Rescan(hash *chainhash.Hash, addrs []btcutil.Address,
+	outPoints []*wire.OutPoint) error {
+
+	// Update addresses.
+	select {
+	case c.rescanUpdate <- addrs:
+	case <-c.quit:
+		return nil
+	}
+
+	// Update outpoints.
+	select {
+	case c.rescanUpdate <- outPoints:
+	case <-c.quit:
+		return nil
+	}
+
+	// Kick off the rescan with a new hash.
```

Comment:
Perhaps this should only be sent if the hash is non-nil? Also the hash is a txid right? (could use a comment pointing that out)

Reply:
The hash is a block. I've updated the argument name to make it clearer, and added a check to return an error if the block hash is nil (no sense in a rescan if you don't have a starting block).